File: backend/crud/locations.py
"""
CRUD operations for Location entities and character-location relationships.

Location operations handle map locations within a world.
Character-location operations track which agents are present at each location via room membership.
"""

import json
import logging
from datetime import datetime, timezone
from typing import List, Optional

import models
import schemas
from database import serialized_write
from sqlalchemy import func
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy.orm import selectinload

logger = logging.getLogger("LocationCRUD")

# System agent groups to exclude when getting "characters" at a location
SYSTEM_AGENT_GROUPS = {"gameplay", "onboarding"}


async def create_location(
    db: AsyncSession,
    world_id: int,
    location: schemas.LocationCreate,
) -> models.Location:
    """Create a new location in a world with gameplay agents added to the room."""
    # Import here to avoid circular dependency
    from crud.worlds import add_gameplay_agents_to_room, get_world

    # Get world to use its owner_id for the location room (scoped to world owner, not global "system")
    world = await get_world(db, world_id)
    if not world:
        raise ValueError(f"World {world_id} not found")

    # Create a room for this location (with world_id set during creation for unique constraint)
    from crud.rooms import create_room

    room = await create_room(
        db,
        schemas.RoomCreate(name=f"Location: {location.display_name or location.name}"),
        owner_id=world.owner_id,  # Use world owner_id to scope rooms per world owner
        world_id=world_id,  # Set world_id during creation for unique constraint
    )

    # Add gameplay agents to the location room
    added_count = await add_gameplay_agents_to_room(db, room.id)
    if added_count == 0:
        logger.warning(f"No gameplay agents found to add to location room {room.id}")

    # Serialize adjacent locations
    adjacent_json = json.dumps(location.adjacent_to) if location.adjacent_to else None

    db_location = models.Location(
        world_id=world_id,
        name=location.name,
        display_name=location.display_name,
        description=location.description,
        position_x=location.position_x,
        position_y=location.position_y,
        adjacent_locations=adjacent_json,
        room_id=room.id,
        is_discovered=location.is_discovered,
    )
    db.add(db_location)

    async with serialized_write():
        await db.commit()

    await db.refresh(db_location)
    return db_location


async def create_new_room_for_location(
    db: AsyncSession,
    location: models.Location,
) -> models.Room:
    """
    Create a fresh room for an existing location (new visit).

    The old room remains in the database for history purposes.
    The location's room_id is updated to point to the new room.

    Args:
        db: Database session
        location: The location to create a new room for

    Returns:
        The newly created Room
    """
    from crud.rooms import create_room
    from crud.worlds import add_gameplay_agents_to_room, get_world

    # Get world for owner_id
    world = await get_world(db, location.world_id)
    if not world:
        raise ValueError(f"World {location.world_id} not found")

    # Create new room with unique name (timestamp ensures uniqueness across visits)
    timestamp = datetime.now(timezone.utc).strftime("%Y%m%d_%H%M%S")
    room = await create_room(
        db,
        schemas.RoomCreate(name=f"Location: {location.display_name or location.name} [{timestamp}]"),
        owner_id=world.owner_id,
        world_id=location.world_id,
    )

    # Add gameplay agents to the new room
    await add_gameplay_agents_to_room(db, room.id)

    # Update location's room_id (old room remains for history)
    location.room_id = room.id

    async with serialized_write():
        await db.commit()

    await db.refresh(location)
    return room


async def get_location(
    db: AsyncSession,
    location_id: int,
) -> Optional[models.Location]:
    """Get a location by ID."""
    result = await db.execute(
        select(models.Location).options(selectinload(models.Location.room)).where(models.Location.id == location_id)
    )
    return result.scalar_one_or_none()


async def get_location_by_name(
    db: AsyncSession,
    world_id: int,
    location_name: str,
) -> Optional[models.Location]:
    """
    Find location by name or display_name (case-insensitive).

    Searches both `name` (folder name) and `display_name` (heading) columns,
    with whitespace/underscore normalization.

    Search order:
    1. Exact match on name
    2. Exact match on display_name
    3. Normalized match (spaces/underscores) on name
    4. Normalized match (spaces/underscores) on display_name
    """
    search_lower = location_name.lower()

    # 1. Try exact match on name (folder name)
    result = await db.execute(
        select(models.Location)
        .options(selectinload(models.Location.room))
        .where(models.Location.world_id == world_id)
        .where(func.lower(models.Location.name) == search_lower)
    )
    location = result.scalar_one_or_none()
    if location:
        return location

    # 2. Try exact match on display_name (heading)
    result = await db.execute(
        select(models.Location)
        .options(selectinload(models.Location.room))
        .where(models.Location.world_id == world_id)
        .where(func.lower(models.Location.display_name) == search_lower)
    )
    location = result.scalar_one_or_none()
    if location:
        return location

    # 3. Try normalized match on name (spaces <-> underscores)
    name_with_underscores = location_name.replace(" ", "_").lower()
    name_with_spaces = location_name.replace("_", " ").lower()

    for normalized in [name_with_underscores, name_with_spaces]:
        if normalized != search_lower:
            result = await db.execute(
                select(models.Location)
                .options(selectinload(models.Location.room))
                .where(models.Location.world_id == world_id)
                .where(func.lower(models.Location.name) == normalized)
            )
            location = result.scalar_one_or_none()
            if location:
                return location

    # 4. Try normalized match on display_name
    for normalized in [name_with_underscores, name_with_spaces]:
        if normalized != search_lower:
            result = await db.execute(
                select(models.Location)
                .options(selectinload(models.Location.room))
                .where(models.Location.world_id == world_id)
                .where(func.lower(models.Location.display_name) == normalized)
            )
            location = result.scalar_one_or_none()
            if location:
                return location

    return None


async def get_locations(
    db: AsyncSession,
    world_id: int,
) -> List[models.Location]:
    """Get all locations in a world."""
    result = await db.execute(
        select(models.Location).where(models.Location.world_id == world_id).order_by(models.Location.id)
    )
    return list(result.scalars().all())


async def delete_location(
    db: AsyncSession,
    location_id: int,
) -> bool:
    """
    Delete a location from the database.

    Also deletes the associated room if it exists.

    Args:
        db: Database session
        location_id: ID of the location to delete

    Returns:
        True if deleted, False if not found
    """
    location = await db.get(models.Location, location_id)
    if not location:
        return False

    room_id = location.room_id

    # Delete the location
    await db.delete(location)

    # Delete associated room if it exists
    if room_id:
        room = await db.get(models.Room, room_id)
        if room:
            await db.delete(room)

    async with serialized_write():
        await db.commit()

    logger.info(f"Deleted location {location_id} (room_id={room_id})")
    return True


async def sync_locations_with_filesystem(
    db: AsyncSession,
    world_id: int,
    world_name: str,
) -> int:
    """
    Sync database locations with filesystem _index.yaml.

    Deletes any locations from the database that no longer exist in the filesystem.

    Args:
        db: Database session
        world_id: ID of the world
        world_name: Name of the world (for filesystem path)

    Returns:
        Number of locations deleted
    """
    from services.location_service import LocationService

    # Get filesystem locations
    fs_locations = LocationService.load_all_locations(world_name)
    fs_location_names = set(fs_locations.keys())

    # Get database locations
    db_locations = await get_locations(db, world_id)

    # Find locations to delete (in DB but not in filesystem)
    deleted_count = 0
    for db_loc in db_locations:
        if db_loc.name not in fs_location_names:
            logger.info(f"Deleting orphaned location '{db_loc.name}' from world '{world_name}'")

            # Clean up room mapping in _state.json
            room_key = f"location:{db_loc.name}"
            LocationService.delete_room_mapping(world_name, room_key)

            # Delete from database
            await delete_location(db, db_loc.id)
            deleted_count += 1

    if deleted_count > 0:
        logger.info(f"Synced locations for world '{world_name}': deleted {deleted_count} orphaned locations")

    return deleted_count


async def update_location_label(
    db: AsyncSession,
    location_id: int,
    label: Optional[str],
) -> Optional[models.Location]:
    """Update a location's user-assigned label."""
    result = await db.execute(select(models.Location).where(models.Location.id == location_id))
    location = result.scalar_one_or_none()

    if not location:
        return None

    location.label = label

    async with serialized_write():
        await db.commit()

    await db.refresh(location)
    return location


async def add_adjacent_location(
    db: AsyncSession,
    location_id: int,
    adjacent_location_id: int,
) -> Optional[models.Location]:
    """Add an adjacent location connection."""
    result = await db.execute(select(models.Location).where(models.Location.id == location_id))
    location = result.scalar_one_or_none()

    if not location:
        return None

    # Parse existing adjacents
    adjacents = json.loads(location.adjacent_locations) if location.adjacent_locations else []

    # Add new adjacent if not already present
    if adjacent_location_id not in adjacents:
        adjacents.append(adjacent_location_id)
        location.adjacent_locations = json.dumps(adjacents)

        async with serialized_write():
            await db.commit()

    await db.refresh(location)
    return location


# =============================================================================
# Character Location Operations (via room membership)
# =============================================================================


async def add_character_to_location(
    db: AsyncSession,
    agent_id: int,
    location_id: int,
) -> bool:
    """
    Add a character (agent) to a location by adding them to the location's room.

    Args:
        db: Database session
        agent_id: ID of the agent to add
        location_id: ID of the location

    Returns:
        True if added successfully, False if location has no room
    """
    from crud.room_agents import add_agent_to_room

    location = await db.get(models.Location, location_id)
    if not location or not location.room_id:
        return False

    await add_agent_to_room(db, location.room_id, agent_id)
    return True


async def remove_character_from_location(
    db: AsyncSession,
    agent_id: int,
    location_id: int,
) -> bool:
    """
    Remove a character (agent) from a location by removing them from the location's room.

    Args:
        db: Database session
        agent_id: ID of the agent to remove
        location_id: ID of the location

    Returns:
        True if removed successfully, False if location has no room
    """
    from crud.room_agents import remove_agent_from_room

    location = await db.get(models.Location, location_id)
    if not location or not location.room_id:
        return False

    await remove_agent_from_room(db, location.room_id, agent_id)
    return True


async def move_character_to_location(
    db: AsyncSession,
    agent_id: int,
    from_location_id: Optional[int],
    to_location_id: int,
) -> bool:
    """
    Move a character from one location to another.

    Args:
        db: Database session
        agent_id: ID of the agent to move
        from_location_id: ID of the source location (can be None)
        to_location_id: ID of the destination location

    Returns:
        True if moved successfully
    """
    # Remove from old location
    if from_location_id:
        await remove_character_from_location(db, agent_id, from_location_id)

    # Add to new location
    return await add_character_to_location(db, agent_id, to_location_id)


async def get_agent_locations_in_world(
    db: AsyncSession,
    agent_id: int,
    world_id: int,
) -> List[models.Location]:
    """
    Get all locations where an agent is present in a specific world.

    Args:
        db: Database session
        agent_id: ID of the agent
        world_id: ID of the world

    Returns:
        List of Location models where the agent is present
    """
    # Query locations in this world where the agent is in the room
    result = await db.execute(
        select(models.Location)
        .where(models.Location.world_id == world_id)
        .options(selectinload(models.Location.room).selectinload(models.Room.agents))
    )
    locations = result.scalars().all()

    # Filter to locations where this agent is in the room
    agent_locations = []
    for location in locations:
        if not location.room:
            continue
        for agent in location.room.agents:
            if agent.id == agent_id:
                agent_locations.append(location)
                break

    return agent_locations


async def get_characters_at_location(
    db: AsyncSession,
    location_id: int,
    exclude_system_agents: bool = True,
) -> List[models.Agent]:
    """
    Get all character agents at a location (agents in the location's room).

    Args:
        db: Database session
        location_id: ID of the location
        exclude_system_agents: If True, excludes gameplay/onboarding agents

    Returns:
        List of Agent models at the location
    """
    # Chain selectinload to load room and its agents in a single query
    # This avoids issues with SQLAlchemy's identity map returning cached objects
    # that don't have the agents relationship loaded
    location = await db.get(
        models.Location,
        location_id,
        options=[selectinload(models.Location.room).selectinload(models.Room.agents)],
    )
    if not location or not location.room:
        return []

    agents = location.room.agents
    if exclude_system_agents:
        agents = [a for a in agents if a.group not in SYSTEM_AGENT_GROUPS]

    return agents


async def get_all_characters_in_world(
    db: AsyncSession,
    world_id: int,
    exclude_system_agents: bool = True,
) -> List[dict]:
    """
    Get all character agents across all locations in a world.

    Args:
        db: Database session
        world_id: ID of the world
        exclude_system_agents: If True, excludes gameplay/onboarding agents

    Returns:
        List of dicts with agent info and their location
    """
    # Get all locations with their rooms and room agents
    result = await db.execute(
        select(models.Location)
        .where(models.Location.world_id == world_id)
        .options(selectinload(models.Location.room).selectinload(models.Room.agents))
    )
    locations = result.scalars().all()

    # Collect unique characters with their location info
    characters = []
    seen_agent_ids = set()

    for location in locations:
        if not location.room:
            continue

        for agent in location.room.agents:
            # Skip if already seen (agent could be in multiple rooms)
            if agent.id in seen_agent_ids:
                continue

            # Skip system agents if requested
            if exclude_system_agents and agent.group in SYSTEM_AGENT_GROUPS:
                continue

            seen_agent_ids.add(agent.id)
            characters.append(
                {
                    "id": agent.id,
                    "name": agent.name,
                    "profile_pic": agent.profile_pic,
                    "in_a_nutshell": agent.in_a_nutshell,
                    "location_id": location.id,
                    "location_name": location.display_name or location.name,
                }
            )

    return characters

File: backend/src/crud/locations.ts
/** CRUD operations for Location entities. */

import { and, asc, eq, isNull, notInArray, or, sql } from 'drizzle-orm'
import type { AnySQLiteColumn } from 'drizzle-orm/sqlite-core'
import type { Db } from '../db'
import {
  agents,
  locations,
  roomAgents,
  rooms,
  worlds,
  type Agent,
  type Location,
  type Room,
} from '../db/schema'
import { getLogger } from '../infrastructure/logging/logger'
import { RoomMappingService } from '../services/room-mapping'
import { addAgentToRoom, createRoom, getAgentsInRoom, removeAgentFromRoom } from './rooms'
import { addGameplayAgentsToRoom } from './worlds'
import { SYSTEM_AGENT_GROUPS } from '../domain/agent'

const logger = getLogger('LocationCRUD')

// A location's room holds the Action_Manager and friends alongside its NPCs, so
// anything user-facing subtracts this set.
export { SYSTEM_AGENT_GROUPS }

export interface LocationWithRoom extends Location {
  room: Room | null
}

export interface LocationCreate {
  name: string
  displayName?: string | null
  description?: string | null
  positionX?: number
  positionY?: number
  /** Ids of locations reachable from here; empty is stored as NULL. */
  adjacentTo?: number[] | null
  isDiscovered?: boolean
  isDraft?: boolean
}

function requireWorldOwner(db: Db, worldId: number): string | null {
  const world = db.select({ ownerId: worlds.ownerId }).from(worlds).where(eq(worlds.id, worldId)).get()
  if (!world) throw new Error(`World ${worldId} not found`)
  return world.ownerId
}

/** `owner_id = ?` that also matches NULL, which SQL equality never does. */
function ownerMatches(ownerId: string | null) {
  return ownerId === null ? isNull(rooms.ownerId) : eq(rooms.ownerId, ownerId)
}

// The `existingRoom` lookup adopts orphans left by older databases. `db` is not
// rebound inside the transaction: `bun:sqlite` scopes one to the *connection*,
// so the helpers below join it and a nested `db.transaction` is a SAVEPOINT.
export function createLocation(db: Db, worldId: number, location: LocationCreate): Location {
  const ownerId = requireWorldOwner(db, worldId)
  const roomName = `Location: ${location.displayName || location.name}`

  return db.transaction(() => {
    const existingRoom = db
      .select()
      .from(rooms)
      .where(and(ownerMatches(ownerId), eq(rooms.name, roomName), eq(rooms.worldId, worldId)))
      .get()

    let roomId: number
    if (existingRoom) {
      logger.info(`Reusing existing room '${roomName}' (id=${existingRoom.id}) for location`)
      roomId = existingRoom.id
    } else {
      // `world_id` at creation, not patched after: it is part of
      // `ux_rooms_owner_name_world`, and the intermediate NULL would collide.
      roomId = createRoom(db, { name: roomName }, ownerId, worldId).id
    }

    if (addGameplayAgentsToRoom(db, roomId) === 0) {
      logger.warning(`No gameplay agents found to add to location room ${roomId}`)
    }

    return db
      .insert(locations)
      .values({
        worldId,
        name: location.name,
        displayName: location.displayName ?? null,
        description: location.description ?? null,
        positionX: location.positionX ?? 0,
        positionY: location.positionY ?? 0,
        adjacentLocations:
          location.adjacentTo && location.adjacentTo.length > 0
            ? JSON.stringify(location.adjacentTo)
            : null,
        roomId,
        isDiscovered: location.isDiscovered ?? true,
        isDraft: location.isDraft ?? false,
      })
      .returning()
      .get()
  })
}

// A fresh visit gets a new room; the previous one is left unreferenced on
// purpose, since it holds the last visit's transcript. The timestamp keeps
// `ux_rooms_owner_name_world` happy — two visits in one second still collide.
export function createNewRoomForLocation(db: Db, location: Location): Room {
  const ownerId = requireWorldOwner(db, location.worldId)
  const stamp = new Date()
    .toISOString()
    .replace(/[-:]/g, '')
    .replace('T', '_')
    .slice(0, 15)

  return db.transaction(() => {
    const room = createRoom(
      db,
      { name: `Location: ${location.displayName || location.name} [${stamp}]` },
      ownerId,
      location.worldId,
    )

    addGameplayAgentsToRoom(db, room.id)
    db.update(locations).set({ roomId: room.id }).where(eq(locations.id, location.id)).run()

    return room
  })
}

export function getLocation(db: Db, locationId: number): LocationWithRoom | null {
  const row = db
    .select({ location: locations, room: rooms })
    .from(locations)
    .leftJoin(rooms, eq(locations.roomId, rooms.id))
    .where(eq(locations.id, locationId))
    .get()

  return row ? { ...row.location, room: row.room } : null
}

/**
 * Find a location by `name` or `display_name`, case-insensitively. Four stages,
 * first hit wins: exact `name`, exact `display_name`, then separator-normalized
 * forms of each. The ordering is load-bearing — a world can hold both an
 * `old_mill` displayed as "Old Mill" and a location literally named "Old Mill",
 * so one OR-ed query, or stage 4 before stage 3, picks the wrong room. The
 * lowercasing is asymmetric: the term in JavaScript (Unicode-aware), the column
 * by SQLite's `lower()` (ASCII-only).
 */
export function getLocationByName(
  db: Db,
  worldId: number,
  locationName: string,
): LocationWithRoom | null {
  const searchLower = locationName.toLowerCase()

  const probe = (column: AnySQLiteColumn, value: string): LocationWithRoom | null => {
    const row = db
      .select({ location: locations, room: rooms })
      .from(locations)
      .leftJoin(rooms, eq(locations.roomId, rooms.id))
      .where(and(eq(locations.worldId, worldId), sql`lower(${column}) = ${value}`))
      .get()
    return row ? { ...row.location, room: row.room } : null
  }

  const exactName = probe(locations.name, searchLower)
  if (exactName) return exactName

  const exactDisplay = probe(locations.displayName, searchLower)
  if (exactDisplay) return exactDisplay

  // `foo bar` <-> `foo_bar`, skipping a variant identical to the original.
  const normalized = [
    locationName.replaceAll(' ', '_').toLowerCase(),
    locationName.replaceAll('_', ' ').toLowerCase(),
  ].filter((candidate) => candidate !== searchLower)

  for (const candidate of normalized) {
    const hit = probe(locations.name, candidate)
    if (hit) return hit
  }

  for (const candidate of normalized) {
    const hit = probe(locations.displayName, candidate)
    if (hit) return hit
  }

  return null
}

export function getLocations(db: Db, worldId: number): Location[] {
  return db
    .select()
    .from(locations)
    .where(eq(locations.worldId, worldId))
    .orderBy(asc(locations.id))
    .all()
}

// Deleting the room is the point: `messages`, `room_agents` and
// `room_agent_sessions` all cascade from it. Rooms from *earlier* visits are
// unreachable from the location row and survive.
export function deleteLocation(db: Db, locationId: number): boolean {
  const location = db
    .select({ roomId: locations.roomId })
    .from(locations)
    .where(eq(locations.id, locationId))
    .get()

  if (!location) return false

  db.transaction(() => {
    db.delete(locations).where(eq(locations.id, locationId)).run()
    if (location.roomId !== null) {
      db.delete(rooms).where(eq(rooms.id, location.roomId)).run()
    }
  })

  logger.info(`Deleted location ${locationId} (room_id=${location.roomId})`)
  return true
}

// Injected rather than imported: the CRUD layer keeps no dependency on services.
export interface LocationFilesystemSync {
  /** Only the keys are read, so a `LocationStorage` satisfies this. */
  loadAllLocations(worldName: string): Record<string, unknown>
  /** Not yet in `services/room-mapping.ts`; a no-op leaks `_state.json` entries. */
  deleteRoomMapping(worldName: string, roomKey: string): void
}

// The filesystem is the source of truth for the map. Matching is on `name` (the
// folder name), never `display_name` — renaming a heading must not delete it.
export function syncLocationsWithFilesystem(
  db: Db,
  worldId: number,
  worldName: string,
  filesystem: LocationFilesystemSync,
): number {
  const onDisk = new Set(Object.keys(filesystem.loadAllLocations(worldName)))

  let deleted = 0
  for (const location of getLocations(db, worldId)) {
    if (onDisk.has(location.name)) continue

    logger.info(`Deleting orphaned location '${location.name}' from world '${worldName}'`)
    // Mapping first: a stale one is harmless if the delete throws, while a
    // mapping onto a deleted room id opens the wrong transcript next turn.
    filesystem.deleteRoomMapping(worldName, RoomMappingService.locationToRoomKey(location.name))
    deleteLocation(db, location.id)
    deleted += 1
  }

  if (deleted > 0) {
    logger.info(`Synced locations for world '${worldName}': deleted ${deleted} orphaned locations`)
  }

  return deleted
}

// `undefined` skips the field, `null` writes NULL — collapsing the two would
// make a nullable column impossible to clear.
export interface LocationUpdate {
  name?: string
  displayName?: string | null
  description?: string | null
  label?: string | null
  positionX?: number | null
  positionY?: number | null
  isDiscovered?: boolean | null
  isDraft?: boolean | null
}

export function updateLocation(
  db: Db,
  locationId: number,
  update: LocationUpdate,
): Location | null {
  const existing = db.select().from(locations).where(eq(locations.id, locationId)).get()
  if (!existing) return null

  const patch: Partial<typeof locations.$inferInsert> = {}
  if (update.name !== undefined) patch.name = update.name
  if (update.displayName !== undefined) patch.displayName = update.displayName
  if (update.description !== undefined) patch.description = update.description
  if (update.label !== undefined) patch.label = update.label
  if (update.positionX !== undefined) patch.positionX = update.positionX
  if (update.positionY !== undefined) patch.positionY = update.positionY
  if (update.isDiscovered !== undefined) patch.isDiscovered = update.isDiscovered
  if (update.isDraft !== undefined) patch.isDraft = update.isDraft

  // Drizzle refuses an empty SET clause.
  if (Object.keys(patch).length === 0) return existing

  const updated = db
    .update(locations)
    .set(patch)
    .where(eq(locations.id, locationId))
    .returning()
    .get()

  logger.info(`Updated location ${locationId}: ${Object.keys(patch).join(', ')}`)
  return updated
}

export function updateLocationLabel(
  db: Db,
  locationId: number,
  label: string | null,
): Location | null {
  return (
    db.update(locations).set({ label }).where(eq(locations.id, locationId)).returning().get() ?? null
  )
}

// Idempotent, and the edge is one-way — a two-way passage needs two calls.
export function addAdjacentLocation(
  db: Db,
  locationId: number,
  adjacentLocationId: number,
): Location | null {
  const existing = db.select().from(locations).where(eq(locations.id, locationId)).get()
  if (!existing) return null

  const adjacents = parseAdjacentIds(existing.adjacentLocations)
  if (adjacents.includes(adjacentLocationId)) return existing

  adjacents.push(adjacentLocationId)
  return db
    .update(locations)
    .set({ adjacentLocations: JSON.stringify(adjacents) })
    .where(eq(locations.id, locationId))
    .returning()
    .get()
}

// `adjacent_locations` holds NULL, `''` or a JSON array depending on which path
// wrote it, and tool handlers write it too — so non-numeric entries are dropped.
function parseAdjacentIds(raw: string | null): number[] {
  if (!raw) return []
  try {
    const parsed: unknown = JSON.parse(raw)
    return Array.isArray(parsed) ? parsed.filter((id): id is number => typeof id === 'number') : []
  } catch {
    return []
  }
}

export interface GetCharactersAtLocationOptions {
  excludeSystemAgents?: boolean
}

// Presence is membership of the location's room — there is no agent→location
// column — so a roomless location has nobody. A NULL `group` survives the
// filter: that is a hand-made character, not machinery.
export function getCharactersAtLocation(
  db: Db,
  locationId: number,
  { excludeSystemAgents = true }: GetCharactersAtLocationOptions = {},
): Agent[] {
  const location = db
    .select({ roomId: locations.roomId })
    .from(locations)
    .where(eq(locations.id, locationId))
    .get()

  if (!location || location.roomId === null) return []

  const present = getAgentsInRoom(db, location.roomId)
  if (!excludeSystemAgents) return present
  return present.filter((agent) => agent.group === null || !SYSTEM_AGENT_GROUPS.has(agent.group))
}

// False means "this location cannot hold characters"; the agent is not checked.
export function addCharacterToLocation(db: Db, agentId: number, locationId: number): boolean {
  const location = db
    .select({ roomId: locations.roomId })
    .from(locations)
    .where(eq(locations.id, locationId))
    .get()

  if (!location || location.roomId === null) return false

  addAgentToRoom(db, location.roomId, agentId)
  return true
}

export function removeCharacterFromLocation(db: Db, agentId: number, locationId: number): boolean {
  const location = db
    .select({ roomId: locations.roomId })
    .from(locations)
    .where(eq(locations.id, locationId))
    .get()

  if (!location || location.roomId === null) return false

  removeAgentFromRoom(db, location.roomId, agentId)
  return true
}

// `fromLocationId` is null for a first placement; the result is the arrival's.
export function moveCharacterToLocation(
  db: Db,
  agentId: number,
  fromLocationId: number | null,
  toLocationId: number,
): boolean {
  if (fromLocationId) {
    removeCharacterFromLocation(db, agentId, fromLocationId)
  }
  return addCharacterToLocation(db, agentId, toLocationId)
}

// One join rather than a per-location loop: `bun:sqlite` is synchronous, so a
// query per location blocks the event loop and stalls in-flight SDK streaming.
export function getAgentLocationsInWorld(db: Db, agentId: number, worldId: number): Location[] {
  return db
    .select({ location: locations })
    .from(locations)
    .innerJoin(roomAgents, eq(roomAgents.roomId, locations.roomId))
    .where(and(eq(locations.worldId, worldId), eq(roomAgents.agentId, agentId)))
    .orderBy(asc(locations.id))
    .all()
    .map((row) => row.location)
}

// The snake_case keys are `GET /worlds/{id}/characters`' response body verbatim
// — renaming one here breaks the frontend's character panel.
export interface WorldCharacter {
  id: number
  name: string
  profile_pic: string | null
  in_a_nutshell: string | null
  location_id: number
  location_name: string
}

export interface GetAllCharactersInWorldOptions {
  excludeSystemAgents?: boolean
}

// De-duplication stays in JavaScript because an agent in two rooms belongs at
// the *first* location by id, and the filter needs its explicit NULL branch:
// `group NOT IN (...)` is NULL for a NULL group, dropping hand-made characters.
export function getAllCharactersInWorld(
  db: Db,
  worldId: number,
  { excludeSystemAgents = true }: GetAllCharactersInWorldOptions = {},
): WorldCharacter[] {
  const notSystem = or(isNull(agents.group), notInArray(agents.group, [...SYSTEM_AGENT_GROUPS]))

  const rows = db
    .select({
      id: agents.id,
      name: agents.name,
      profilePic: agents.profilePic,
      inANutshell: agents.inANutshell,
      locationId: locations.id,
      displayName: locations.displayName,
      locationName: locations.name,
    })
    .from(locations)
    .innerJoin(roomAgents, eq(roomAgents.roomId, locations.roomId))
    .innerJoin(agents, eq(agents.id, roomAgents.agentId))
    .where(
      excludeSystemAgents
        ? and(eq(locations.worldId, worldId), notSystem)
        : eq(locations.worldId, worldId),
    )
    .orderBy(asc(locations.id))
    .all()

  const characters: WorldCharacter[] = []
  const seen = new Set<number>()

  for (const row of rows) {
    if (seen.has(row.id)) continue
    seen.add(row.id)
    characters.push({
      id: row.id,
      name: row.name,
      profile_pic: row.profilePic,
      in_a_nutshell: row.inANutshell,
      location_id: row.locationId,
      location_name: row.displayName || row.locationName,
    })
  }

  return characters
}
